### ISSUE TITLE
Handling wildcard CNAME loops

### ISSUE THREAD

**Issue by SivaKesava1** [coredns/coredns#4378]
**What would you like to be added**:
This is a rare error case, so I am putting it under enhancement.  Consider the following zone file:

|  	|  	|  	|  	|
|-	|-	|-	|-	|
|example.com. | 500 | SOA | ns1.outside.com. root.example.com. 3 604800 86400 2419200 604800|
|example.com. | 500 | NS | ns1.outside.com.|
|a.example.com|500|CNAME|b.example.com.|
|b.example.com|500|CNAME|a.example.com.|
|*.foo.example.com|500|CNAME|bar.foo.example.com|

Coredns handles very well when the query is either for `a.example.com` or `b.example.com`.  A small issue arises when the query is for `something.foo.example.com.` The server does not respond at all, and also, there is no log that this query is received. The `dig` response is a `connection timed out; no servers could be reached.` It could be handled the same way as above (even self-loops are handled well by Coredns).

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "[coredns/coredns] Handling wildcard..." ]
>What would you like to be added:
>This is a rare error case, so I am putting it under enhancement. Consider the
>following zone file:
>
>example.com.      500 SOA   ns1.outside.com. root.example.com. 3 604800 86400
>                            2419200 604800
>example.com.      500 NS    ns1.outside.com.
>a.example.com     500 CNAME b.example.com.
>b.example.com     500 CNAME a.example.com.
>*.foo.example.com 500 CNAME bar.foo.example.com
>
>Coredns handles very well when the query is either for a.example.com or
>b.example.com. A small issue arises when the query is for
>something.foo.example.com. The server does not respond at all, and also, there
>is no log that this query is received. The dig response is a connection timed
>out; no servers could be reached. It could be handled the same way as above
>(even self-loops are handled well by Coredns).

can't replicate with master: zone is loaded:

~~~
./coredns -dns.port=1053
example.com.:1053
CoreDNS-1.8.1
linux/amd64, go1.15.4,
~~~

~~~
% dig @localhost -p 1053 something.foo.example.com

; <<>> DiG 9.16.1-Ubuntu <<>> @localhost -p 1053 something.foo.example.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 27880
;; flags: qr aa rd; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
; COOKIE: 0379272418df2015 (echoed)
;; QUESTION SECTION:
;something.foo.example.com.	IN	A

;; AUTHORITY SECTION:
example.com.		500	IN	SOA	ns1.outside.com. root.example.com. 3 604800 86400 2419200 604800

;; Query time: 0 msec
;; SERVER: 127.0.0.1#1053(127.0.0.1)
;; WHEN: di jan 12 14:14:09 CET 2021
;; MSG SIZE  rcvd: 144
~~~

**Comment by SivaKesava1** [comment]
Sorry, my bad, the last '.' was missing in the records. Coredns would have taken them as relative names instead of as FQDN. 

Consider the same one but properly formatted:

example.com.	500	IN	SOA	ns1.outside.com. root.example.com. 3 604800 86400 2419200 604800
example.com.	500	IN	NS	ns1.outside.com.
a.example.com.	500	IN	CNAME	b.example.com.
b.example.com.	500	IN	CNAME	a.example.com.
*.foo.example.com.	500	IN	CNAME	bar.foo.example.com.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Handling wild..." ]
>Sorry, my bad, the last '.' was missing in the records. Coredns would have
>taken them as relative names instead of as FQDN.

LOL - I missed that too. Get the loop going

**Comment by miekg** [comment]
See https://github.com/coredns/coredns/pull/4387 which fixes. Returns bogus answers, but that's what you ask for by creating such a zone :-) 
thanks for reporting this!

**Comment by miekg** [comment]
hmm, looking more closely at the testcase I see that's doing the right thing. But running a fullblown server is adding records. Have to dig slightly deeper to understand why

**Comment by miekg** [comment]
closed via #4387
